pkg/apis/core/validation/validation.go

```diff
@@ -4154,12 +4163,19 @@ func ValidateServiceExternalTrafficFieldsCombination(service *core.Service) fiel
 func ValidateServiceUpdate(service, oldService *core.Service) field.ErrorList {
 	allErrs := ValidateObjectMetaUpdate(&service.ObjectMeta, &oldService.ObjectMeta, field.NewPath("metadata"))
 
-	// ClusterIP should be immutable for services using it (every type other than ExternalName)
+	// ClusterIP and IPFamily should be immutable for services using it (every type other than ExternalName)
 	// which do not have ClusterIP assigned yet (empty string value)
 	if service.Spec.Type != core.ServiceTypeExternalName {
 		if oldService.Spec.Type != core.ServiceTypeExternalName && oldService.Spec.ClusterIP != "" {
 			allErrs = append(allErrs, ValidateImmutableField(service.Spec.ClusterIP, oldService.Spec.ClusterIP, field.NewPath("spec", "clusterIP"))...)
 		}
+		// notes:
+		// we drop the IPFamily field when the Dualstack gate is off.
+		// once the gate is on, we start assigning default ipfamily according to cluster settings. in other words
+		// though the field is immutable, we allow (onetime) change from nil==> to value
+		if oldService.Spec.IPFamily != nil {
```

Comment:
If we have the defaulting in place, we should not ever observe this as nil when the gate is enabled.

When the gate is disabled and oldObj.ipFamily was nil, we should have blanked newObj in REST already, so no need to nil check.

When the gate is disabled and oldObj.ipFamily was non-nil, no need to nil-check.


So...no need to nil-check?

Reply:
as an existing cluster moves from single stack to dualstack, the field will be set by ipallocator on new service, but it will not be set on old service (this is ok and expected). What is not ok is setting the field to a value different to `old service non nil IPFamily` (irrespective of dual stack or not).

Reply:
The read operation will default the value from clusterIP, so we can check immutable without testing for nil.

If the old value was non-nil, we will not clear it (dropDisabled is explicit about that) in strategy, so we can check immutability without testing nil.

If the old value was nil, the new value must also be nil, which would be enforced by strategy, but checking it here is fine.

Reply:
So after actually trying this, this broke old clients (they will send nil family on an update). #86895 has a new test which shows this behavior.  Basically, we need to default ipfamily on every create and update, and allow nil ipfamily on update.